File: java/src/test/java/kkurczewski/deser/greet/GreetingControllerTest.java
package kkurczewski.deser.greet;

import kkurczewski.deser.greet.dto.Greeting;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest;
import org.springframework.test.web.reactive.server.WebTestClient;

import java.net.URI;

@WebFluxTest(GreetingController.class)
class GreetingControllerTest {

    @Autowired
    private WebTestClient webClient;

    @Test
    void shouldEchoRequest() {
        Greeting greeting = new Greeting("Ahoj");

        webClient
                .post()
                .uri(URI.create("/greet"))
                .bodyValue(greeting)
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Greeting.class)
                .isEqualTo(greeting)
                .returnResult();
    }

    @Test
    void validateJsonFormat() {
        Greeting greeting = new Greeting("Ahoj");

        webClient
                .post()
                .uri(URI.create("/greet"))
                .bodyValue(greeting)
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody()
                .json(normalize("{'greeting':'Ahoj'}"));
    }

    private String normalize(String json) {
        return json.replaceAll("'", "\"");
    }

}

File: kotlin/src/test/kotlin/pl/kkurczewski/deser/greet/GreetingControllerTest.kt
package pl.kkurczewski.deser.greet

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest
import org.springframework.test.web.reactive.server.WebTestClient
import pl.kkurczewski.deser.greet.dto.Greeting
import java.net.URI

@WebFluxTest(GreetingController::class)
class GreetingControllerTest(@Autowired val webClient: WebTestClient) {

    @Test
    fun shouldEchoRequest() {
        val greeting = Greeting("Ahoj")
        val body = webClient
                .post()
                .uri(URI.create("/greet"))
                .bodyValue(greeting)
                .exchange()
                .expectStatus()
                .is2xxSuccessful
                .expectBody(Greeting::class.java)
                .returnResult()
                .responseBody

        assertThat(body).isEqualTo(greeting)
    }

    @Test
    fun validateJsonFormat() {
        val greeting = Greeting("Ahoj")
        webClient
                .post()
                .uri(URI.create("/greet"))
                .bodyValue(greeting)
                .exchange()
                .expectStatus()
                .is2xxSuccessful
                .expectBody()
                .json(normalize("{'greeting':'Ahoj'}"))
    }

    private fun normalize(json: String): String {
        return json.replace("'".toRegex(), "\"")
    }
}